core/src/main/scala/caliban/execution/Executor.scala

```diff
@@ -106,23 +172,53 @@ object Executor {
         }
       }
 
-      def reduceListStep(steps: List[Step[R]]) = {
-        var i         = 0
-        val lb        = ListBuffer.empty[ReducedStep[R]]
-        val nil       = Nil
-        var remaining = steps
-        while (remaining ne nil) {
-          lb addOne reduceStep(remaining.head, currentField, arguments, PathValue.Index(i) :: path)
-          i += 1
-          remaining = remaining.tail
+      def reduceListStep(steps: List[Step[R]]): ReducedStep[R] = {
+        def reduceMixed(head: ReducedStep[R], remaining0: List[Step[R]]): ReducedStep[R] = {
+          var i         = 1
+          val nil       = Nil
+          val lb        = ListBuffer.empty[ReducedStep[R]]
+          lb addOne head
+          var remaining = remaining0
+          while (remaining ne nil) {
+            val step = reduceStep(remaining.head, currentField, arguments, PathValue.Index(i) :: path)
+            lb addOne step
+            i += 1
+            remaining = remaining.tail
+          }
+          ReducedStep.ListStep(
+            lb.result(),
+            Types.listOf(currentField.fieldType) match {
+              case Some(tpe) => tpe.isNullable
+              case None      => false
+            }
+          )
         }
-        reduceList(
-          lb.result(),
-          Types.listOf(currentField.fieldType) match {
-            case Some(tpe) => tpe.isNullable
-            case None      => false
+
+        def reducePures(head: PureStep, remaining0: List[Step[R]]): ReducedStep[R] = {
+          var i         = 1
+          val nil       = Nil
+          val lb        = ListBuffer.empty[ResponseValue]
+          var remaining = remaining0
+          lb addOne head.value
+          while (remaining ne nil) {
+            val step = reduceStep(remaining.head, currentField, arguments, PathValue.Index(i) :: path)
+            lb addOne step.asInstanceOf[PureStep].value
+            i += 1
+            remaining = remaining.tail
           }
-        )
+          PureStep(ListValue(lb.result()))
+        }
+
+        if (steps.isEmpty) PureStep(ListValue(Nil))
+        else {
+          val step = reduceStep(steps.head, currentField, arguments, PathValue.Index(0) :: path)
+          if (step.isPure) {
+            // In 99.99% of the cases, if the head is pure, all the other elements will be pure as well but we catch that error just in case
+            // NOTE: Our entire test suite passes without catching the error
+            try reducePures(step.asInstanceOf[PureStep], steps.tail)
```

Comment:
For the first step in the list, we already do the `isInstanceOf` check via the `isPure` method. The `try` block is for catching any cast errors in extremely unlikely cases where a list contains a mixture of pure and effectful steps. Because this is a very unlikely case (a list will usually contain either only pure or only effectful steps), we avoid type-checking every element in the list other than the first one